slime/utils/ppo_utils.py

```diff
@@ -119,3 +121,111 @@ def get_grpo_returns(
+    for i in range(len(rewards)):
+        r = -kl_coef * kl[i]
+        assert loss_masks[i].sum() > 0, f"Sequence at index {i} is fully masked..."
+
+        # Find the index of the last valid token to add the final task reward.
+        last_response_idx = loss_masks[i].nonzero(as_tuple=True)[0][-1]
+        r[last_response_idx] += rewards[i]
+        token_level_rewards.append(r)
+
+    # Calculate discounted returns per sequence
+    returns = []
+    for r_i in token_level_rewards:
+        seq_len = len(r_i)
+        returns_i = torch.zeros_like(r_i)
+        running_return = 0.0
+        for t in reversed(range(seq_len)):
+            # G_t = r_t + gamma * G_{t+1}
+            running_return = r_i[t] + gamma * running_return
+            returns_i[t] = running_return
+        returns.append(returns_i)
+
+    all_returns = torch.cat(returns)
+    all_masks = torch.cat(loss_masks)
+
+    whitened_advs_flat = distributed_masked_whiten(all_returns, all_masks, shift_mean=True)
+    advantages = list(torch.split(whitened_advs_flat, response_lengths))
+
+    return advantages, returns
+
+def get_reinforce_plus_plus_baseline_advantages(
+    rewards: torch.Tensor,
+    kl: List[torch.Tensor],
+    loss_masks: Optional[List[torch.Tensor]],
+    response_lengths: List[int],
+    kl_coef: float,
+) -> List[torch.Tensor]:
+    """
+    Calculates the final advantages for the REINFORCE++-baseline algorithm.
+
+    This process involves two main steps:
+    1. Broadcasting the scalar (reward - group_baseline) to each token.
+    2. Applying a *distributed* global whitening (normalization) across all 
+       advantages in the data-parallel group.
+
+    Args:
+        rewards (Tensor): A tensor of scalar rewards, where the group-wise
+                                baseline has already been subtracted.
+        kl (list[Tensor]): A list of per-token KL divergence tensors. Used to
+                                 get the shape for broadcasting.
+        loss_masks (list[Tensor] | None): A list of per-token loss masks. If None,
+                                          it's assumed all tokens are active.
+        response_lengths (list[int]): A list of sequence lengths, required for
+                                      splitting the whitened tensor back.
+
+    Returns:
+        list[Tensor]: A list of tensors containing the final, whitened advantages.
+    """
+    # Broadcast to get unwhitened advantages
+    unwhitened_advantages = [
+        torch.ones_like(kl_tensor) * reward_val - kl_coef * kl_tensor
+        for kl_tensor, reward_val in zip(kl, rewards)
+    ]
+    # Concatenate tensors for a global operation
+    if loss_masks is None:
+        loss_masks = [
+            torch.ones_like(adv) for adv in unwhitened_advantages
+        ]
+
+    all_advs = torch.cat(unwhitened_advantages)
+    all_masks = torch.cat(loss_masks)
+    
+    whitened_advs_flat = distributed_masked_whiten(all_advs, all_masks, shift_mean=True)
```

Comment:
Can we move the `distributed_masked_whiten` to the `NotImplementedError` in `if args.normalize_advantages:`?